Resolve August integration makes too many requests and hits rate limits

bdraco opened the pull request:
Significantly reduce the number of api calls
that the august integration makes

Lock and door status now use the get_lock_status call with
door_status=True to avoid multiple calls to the same endpoint

Lock and door status are now updated ever 10 seconds instead
of every 5 seconds

Lock detail (battery level, etc) are now only updated every
1800s since its unlikely to change even of that period

A bug that caused the activities endpoint to be queried with
the exact same request once per lock or doorbell instead of
once per house has been resolved (self._house_ids had
duplicate ids)

<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Breaking change
<!--
  If your PR contains a breaking change for existing users, it is important
  to tell them what breaks, how to make it work again and why we did this.
  This piece of text is published with the release notes, so it helps if you
  write it towards our users, not us.
  Note: Remove this section if this PR is NOT a breaking change.
-->


## Proposed change
<!-- 
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->


## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box! 
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [x] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Example entry for `configuration.yaml`:
<!--
  Supplying a configuration snippet, makes it easier for a maintainer to test
  your PR. Furthermore, for new integrations, it gives an impression of how
  the configuration would look like.
  Note: Remove this section if this PR does not have an example entry.
-->

```yaml
# Example configuration.yaml

```

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #31472
- This PR is related to issue: 
- Link to documentation pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [x] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [ ] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

homeassistant commented:
Hi @bdraco, 

It seems you haven't yet signed a CLA. Please do so [here](https://home-assistant.io/developers/cla_sign_start/?pr=home-assistant&#x2F;home-assistant#31558).

Once you do that we will be able to review and accept this pull request.

Thanks!

dannyk81 commented on homeassistant/components/august/__init__.py:
```diff
@@ -284,58 +294,51 @@ def get_door_state(self, lock_id):
             except RequestException as ex:
                 _LOGGER.error(
-                    "Request error trying to retrieve door status for %s. %s",
+                    "Request error trying to retrieve locka dn door status for %s. %s",
```
looks like a small typo

```suggestion
                    "Request error trying to retrieve lock and door status for %s. %s",
```

bdraco commented:
> Thanks for this ! Would it make sense to rework August to use the Data Update Coordinator helper that we have? ([docs](https://developers.home-assistant.io/docs/en/integration_fetching_data.html))

The likely makes more sense here, however its a quite an increase in scope so I'm not sure if I can find the time to do it in the next few days. @balloob Do you have an integration in mind that is a good example of how to do this well?

balloob commented:
Philips hue. Which also inspired the docs.

I'll review this tomorrow, get this merged first.

bdraco commented:
Assuming py-august doesn’t take the pr I mentioned in the original issue, it’s likely best to merge this with an update interval of 60s instead of 10s

bdraco commented:
Might be able to use the activity endpoint to significantly reduce the number of lock queries since the unlock/lock events are in there as well

bdraco commented:
I'm going to assume the the py-august author isn't available at the moment and rework this to handle the 429 in the component for now

bdraco commented:
@balloob should be ready for a first pass review.

balloob commented on homeassistant/components/august/__init__.py:
```diff
@@ -170,6 +181,39 @@ def close_http_session(event):
+        # version so lets try to avoid the 429 up front
+        kwargs["headers"][
+            "User-Agent"
+        ] = "August/2019.12.16.4708 CFNetwork/1121.2.2 Darwin/19.3.0"
```
This should not be in Home Assistant. This should be in the lib.

MartinHjelmare replied:
We won't accept this code in home assistant. If the library maintainer is unresponsive for a reasonable time, eg more than one month, and theres no other activity in the library, the library can be forked.

Please try to contact the maintainer to establish a plan going forward. Maybe the maintainer is willing to transfer ownership of the library?

bdraco replied:
@MartinHjelmare The py-august maintainer has responded and released a new version.  PR resubmitted. 

This should resolve @balloob 's review comments

bdraco reviewed (commented):
I'm not sure when the original py-author will be available, however the PR is submitted here that would allow removing the 429 handling from this PR

https://github.com/snjoetw/py-august/pull/24

bdraco commented:
py-august upstream has been updated. I'm reworking this now

bdraco commented:
Looks like there are some lint issues that still need to be worked out

homeassistant commented:
Hi @lamandrice, 

It seems you haven't yet signed a CLA. Please do so [here](https://home-assistant.io/developers/cla_sign_start/?pr=home-assistant&#x2F;home-assistant#31558).

Once you do that we will be able to review and accept this pull request.

Thanks!

bdraco committed:
Resolve August integration makes too many requests and hits rate limits

Significantly reduce the number of api calls
that the august integration makes

Lock and door status now use the get_lock_status call with
door_status=True to avoid multiple calls to the same endpoint

Lock and door status are now updated ever 10 seconds instead
of every 5 seconds

Lock detail (battery level, etc) are now only updated every
1800s since its unlikely to change even of that period

A bug that caused the activities endpoint to be queried with
the exact same request once per lock or doorbell instead of
once per house has been resolved (self._house_ids had
duplicate ids)

This change bumps py-august to 0.8.1 which handles
429 http responses and will retry

bdraco commented:
hmm
ci failed with
```
An unexpected error has occurred: CalledProcessError: command: ('/usr/bin/git', 'fetch', 'origin', '--tags')
return code: 128
expected return code: 0
stdout: (none)
stderr:
    error: RPC failed; curl 56 GnuTLS recv error (-54): Error in the pull function.
    fatal: The remote end hung up unexpectedly
    fatal: early EOF
    fatal: index-pack failed
```

bdraco commented:
> hmm
> ci failed with
> 
> ```
> An unexpected error has occurred: CalledProcessError: command: ('/usr/bin/git', 'fetch', 'origin', '--tags')
> return code: 128
> expected return code: 0
> stdout: (none)
> stderr:
>     error: RPC failed; curl 56 GnuTLS recv error (-54): Error in the pull function.
>     fatal: The remote end hung up unexpectedly
>     fatal: early EOF
>     fatal: index-pack failed
> ```

Failure was transient .. bumping the sha and repushing solved it.

dannyk81 commented on homeassistant/components/august/manifest.json:
```diff
@@ -2,7 +2,7 @@
   "name": "August",
   "documentation": "https://www.home-assistant.io/integrations/august",
-  "requirements": ["py-august==0.7.0"],
+  "requirements": ["py-august==0.8.1"],
```
@bdraco looks like py-august 0.10.0 eas released shortly after 0.8
1 (with token auto renewal), might be relevant to bump it here. wdyt?

bdraco replied:
@dannyk81 
While 0.10.0 makes thing nicer with auto token refresh, When I tested 0.10.0, I found it has a blocking bug:
https://github.com/snjoetw/py-august/pull/25

Once it is fixed, it can be done in a new PR.  Ideally this PR focuses on getting the integration working again and we can do improvements in a followup.

MartinHjelmare commented on homeassistant/components/august/__init__.py:
```diff
@@ -180,7 +187,10 @@ def __init__(self, hass, api, access_token):
-        self._house_ids = [d.house_id for d in self._doorbells + self._locks]
+        self._house_ids = []
+        for device in self._doorbells + self._locks:
+            if device.house_id not in self._house_ids:
```
If we want this container to have unique items we can use a set instead of a list. Then we can simplify here since a set by default won't add duplicate items.

bdraco replied:
will do

MartinHjelmare reviewed (approved):
Looks good!
